### PR TITLE
JS OOP Practice 1 - Eric George

### PR DESCRIPTION
**Pull request by emg187**

### CODE REVIEW

File: exercises/bank.js [hunk 1]
```diff
@@ -1,37 +1,67 @@
 
 
 class Bank{
-	constructor(){
+	constructor(bankName){
```
Comment (BrettAlbright) [inline exercises/bank.js:4]: One of the most important things to consider when building an application in OOP is the functionality that each of the objects you are creating are going to need. In this particular example you are being asked to create a “bank” object, and you should notice as you look at the methods included in this Class constructor, that they are all related to the functionality a bank would need to have. For example, the object can make accounts, check to see if accounts exist, remove accounts, add money to accounts, and withdraw money from accounts. I’m sure you are thinking right now what other features could be helpful. Perhaps you would want to be able to update the name of an account, or add an interest rate… maybe even store the name of account holders? These are the kinds of things you should be considering when designing your objects, so make sure to take a moment when planning object structure so that a vital piece of functionality is not missed!

File: exercises/bankaccount.js [hunk 2]
```diff
@@ -2,17 +2,36 @@
 
 class Account{
```
Comment (BrettAlbright) [inline exercises/bankaccount.js:3]: Like the bank object, the Account object has certain methods that it needs to perform its job. In this instance, we have methods that add money to an account, remove money from an account, and return the total amount of money in an account.Take a moment and consider what other methods you might want to add in you were to build this object out further. Thought experiments like this are good for practicing the conceptualization of objects.

File: exercises/basiccalcobject.js [hunk 3]
```diff
@@ -1,18 +1,44 @@
 
 class Calculator{
 	constructor(){
```
Comment (BrettAlbright) [inline exercises/basiccalcobject.js:3]: One of the most important aspects of the constructor in a class constructor is that it serves as the global space for the object. In this example of a calculator object, the constructor is acting as a storage space for the operator that will eventually be passed in, as well as the numbers that will be part of the calculation. The constructor is incredibly important in this way, because it allows you to store information in an object that is readily available to all the methods attached to that object. This allows you to do two things, organize the data the object methods need to function, and to centralize data related to the object (for example the “name” of a person object, or the numbers/operator data present in this example). Organization is extremely important in programming, as I am sure you have noticed, and constructors are very helpful in this regard.

File: exercises/randomgenerator.js [hunk 4]
```diff
@@ -2,21 +2,34 @@
 
 class RandomGenerator{
 	constructor(min, max){
```
Comment (BrettAlbright) [inline exercises/randomgenerator.js:4]: The “class constructor” of an object is the section which assigns "values" to different "keys" based on what is passed in when the object is instantiated. For example, in this prototype, when a new "randomgenerator" object is created the “min” and “max” values are passed in and assigned to the object that is created. One could imagine that if you we making the constructor for a "student" object that it would contain such key/value pairs as "name", "age", "grade", and maybe even "attendance". If this was the case, when the new object was created and values were passed in for these "keys" they would be assigned to the new object. Great work on this!

File: exercises/card.js [hunk 5]
```diff
@@ -14,27 +14,45 @@ class Card{
 
 class Deck{
 	constructor(){
+		this.cards = [];
 		//takes in nothing
 		//make storage for your card objects
 	}
-	addCard( suit, faceValue ){
+	addCard(suit, faceValue){
+		var card = new Card(suit, faceValue);
```
Comment (BrettAlbright) [inline exercises/card.js:22]: Here is a great example of inter-object-creation. It is important for a deck of cards to have the ability to create and store a deck of cards, so the needed functionality is achieved in this example by having the “addCard” method create and then store the cards. It is common to have objects create other objects, (as you had the bank object create accounts to be stored in the bank’s global “accounts” variable), so it is something that should be kept in mind when designing your class constructor functionality.

File: exercises/randomgenerator.js [hunk 6]
```diff
@@ -2,21 +2,34 @@
 
 class RandomGenerator{
 	constructor(min, max){ 
-		this.min = min;
-		this.max = max;
+		if (typeof min === "undefined" || typeof max === "undefined"){
```
Comment (BrettAlbright) [inline exercises/randomgenerator.js:5]: Another common thing in object creation, (and programming at large) is having default values available for functions to use if no other values are available. In this example of a random number generator, we want the generator to be able to generate a number between the min and max number we supply to it, or to generate a number between 1 and 10 if we do not supply the number. The approach used here is fine, but we can achieve the same functionality in a different way using default values. For example, instead of using a conditional check to see if the “min” and “max” variables are “undefined, you could use a “default or || statement” to check if there is a value to min, and if not, to assign a default value: 

```javascript
    constructor(min, max){
this.min = min || 1
this.max = max || 10;
this.random = null;
}
```
This example will assign to “this.min” what was passed in if a value was passed in, or a 1 if no value was passed in. But, there is an issue with this solution. The issue is that if a “falsey” value is passed in, (like the number zero), the “or” || statement will always choose the default value, (in this case 1). This is a very important thing to consider when using “or” || statements, as any falsy value will cause them to use the default value. We can fix this nature in this example by creating a conditional to account for the “zero” being passed in. for example:

```javascript
if(min === 0) {
this.min = min
} else {
this.min = min || 0
this.max = max || 10
```
Now if zero is passed in, the the conditional will handle it.

File: exercises/bank.js [hunk 7]
```diff
@@ -1,37 +1,67 @@
...
+		return newAccount;
 		//makes an account
 		//takes in the account number of an account
 		//checks if the account number already exists in the account storage object in the constructor
 			//if already exists, returns false
 		//otherwise makes an new Account object, storing it into the storage object in the constructor
 		//returns the Account object that was constructed
 	}
-	checkForAccount(  ){
+	checkForAccount(accountNumber){
```
Comment (BrettAlbright) [inline exercises/bank.js:24]: Functions that are included in a "class constructor" are called methods, and will be attached to any object that is created by the "class constructor". This is incredibly useful for easily adding functionality to objects with the ease of an assembly line. Imagine creating a "person class constructor" that needed to have methods such as "eating", "drinking", and "studying". By doing this, every student that was created by the "class constructor" would be able to eat, drink, and study (I would say "sleep" would be a good method, but most students don't get to do that). 
The utility of this is easy to understand, as it allows code to be easily replicated in each new object simply by calling the "class constructor" Great Job!

**Review by Zhiwutian — COMMENTED** [review]
Hey Eric, One of the most difficult aspects of switching from procedural programming to OOP is the conceptualization of the objects that will be required in order to complete the functionality of the app in question. What this means in plain English is that in OOP you are required to plan out the objects you will need for your app, and the functionality each of the objects will need to have to complete their work. This can be extremely frustrating in the beginning, (and longer, as many things in programming maintain their playful frustrations no matter how long we program), but it is important that these struggles you will face are all a part of the educational process, and that the more things you create in OOP, the more easier it will become to understand.
